### DISCUSSION TITLE
Creating dashboards with only a GPX file...

### DISCUSSION THREAD

**Discussion by time4tea** [time4tea/gopro-dashboard-overlay#66] [Ideas]
This has been asked for a few times - maybe worth a discussion about how people would use it?...

**Comment by michaelochs** [comment]
I am editing aviation videos in FinalCut Pro with three to four camera angles (a mix of GoPros and other cameras). I would love to have this because I would then be able to create the overlay once, place it above every outside camera and then just use the Multicam editing workflow in FinalCut to quickly get my videos done.

The GPX file would come from the airplane's GPS, so it would be super accurate but obviously would have different start / end points than the camera footage. Since I would use the same overlay with multiple cameras, the syncing would need to be done manually anyway however.

Another workflow that I would like to experiment with is to keep the overlay on its separate track so that I can fade it in and out whenever I want to emphasize the current location / speed. This would also be possible with the overlay as its separate video.

Last but not least I think it's important to point out that for this kind of footage it really doesn't matter if the GPX track is a few seconds behind since you are covering a fairly large area and you can't really pin point the exact location on the ground anyway aside from takeoff and landing.

**Reply by KyleGW** [reply]
I have the same workflow and usecase editing flight videos with Davinci resolve. I currently use a combination of stitching together the original video files with no overlay, add overlays to each individual clip from each of the three gopros, and then meshing them together in Davinci after I create multicam clips from all of the originals as well as the copies with the overlays. I also create multiple copies of the same overlay based on what zoom level I want for the map at various points of the video. This takes up a ton of disk space. I also have to first join all of the videos from a single camera together in order to get a full track of the flight, otherwise you get just the section of the track from that particular video clip.

**Comment by time4tea** [comment]
Thanks! I am reading these..!

**Comment by ptanov** [comment]
Hope that #67 will works

**Comment by time4tea** [comment]
There is a branch https://github.com/time4tea/gopro-dashboard-overlay/tree/36-overlay-use-just-gpx which based on some great work by @ptanov might work for creating GPX-Only overlays. The PR that was submitted did in fact work, so if I've messed it up, that's on me.

If you are comfortable using this software from a branch, please do give it a try.

**Comment by time4tea** [comment]
Experimental support for this is now in 0.59.0 - thanks @ptanov

**Comment by time4tea** [comment]
If anybody had success or not-success in using this it would be super to get any feedback!

**Reply by ptanov** [reply]
Hi,
I did some tests but probably I'm missing something. Currently I don't have time to debug it (sorry :( ). I'm attaching gpx file (c.gpx [c.gpx.zip](https://github.com/time4tea/gopro-dashboard-overlay/files/9676482/c.gpx.zip) ), video file is random 20 seconds non-gopro video that is `touch -d "2022-07-26T14:14:58.000Z" nogopro.MP4` so it matches the period of the GPX file (I can provide it if you need it):

Test 1) Generate overlay on top of the original video from gpx and non-gopro video (crop only segment that matches the video)
Test 2) Generate overlay without video (using GPX)
Test 3) Generate overlay only for the part of the non-gopro video (crop only segment that matches the video)

## Generate overlay on top of the original video from gpx and non-gopro video (crop only segment that matches the video)

### On my branch (with patch applied):

```bash
pkill ffmpeg ; pkill mpv ; time venv/bin/gopro-dashboard.py --overlay-size 1920x800 --video-time end-mtime --profile overlay --layout xml --layout-xml  layout.xml --gpx c.gpx nogopro.MP4 result.mov && mpv result.mov

real	0m19,320s
user	2m1,489s
sys	0m2,132s
```

### Using venv/bin/pip3.8 install gopro-overlay

```bash
pkill ffmpeg ; pkill mpv ; time venv/bin/gopro-dashboard.py --use-gpx-only --overlay-size 1920x800 --video-time-end file-modified --profile overlay --layout xml --layout-xml  layout.xml --gpx c.gpx nogopro.MP4 result.mov && mpv result.mov

real	0m19,606s
user	2m1,569s
sys	0m1,886s
```

The result is as expected

## Generate overlay without video (using GPX)

### On my branch (with patch applied):

```bash
pkill ffmpeg ; pkill mpv ; time venv/bin/gopro-dashboard.py --overlay-size 1920x800 --profile overlay --generate overlay --layout xml --layout-xml  layout.xml --gpx c.gpx - result.mov && mpv result.mov


real	0m27,405s
user	1m45,319s
sys	0m2,950s
```

### Using venv/bin/pip3.8 install gopro-overlay

```bash
pkill ffmpeg ; pkill mpv ; time venv/bin/gopro-dashboard.py --use-gpx-only --overlay-size 1920x800 --profile overlay --layout xml --layout-xml  layout.xml --gpx c.gpx result.mov && mpv result.mov

Starting gopro-dashboard version 0.59.0
Timer(loading timeseries - Called: 1, Total: 0.00926, Avg: 0.00926, Rate: 107.97)
Timer(program - Called: 1, Total: 0.00931, Avg: 0.00931, Rate: 107.46)
Traceback (most recent call last):
  File "venv/bin/gopro-dashboard.py", line 119, in <module>
    frame_meta = timeseries_to_framemeta(gpx.load_timeseries(args.gpx, units), units, start_date=start_date, duration=duration)
  File "venv/lib/python3.8/site-packages/gopro_overlay/framemeta_gpx.py", line 71, in timeseries_to_framemeta
    end_date = datetime.datetime.max.astimezone(datetime.timezone.utc)
ValueError: year 10000 is out of range
```

Sorry, don't have time to check what happens (probably you can reproduce it easily using c.gpx)

## Generate overlay only for the part of the non-gopro video (crop only segment that matches the video)

### On my branch (with patch applied):

```bash
pkill ffmpeg ; pkill mpv ; time venv/bin/gopro-dashboard.py --overlay-size 1920x800 --video-time end-mtime --profile overlay --generate overlay --layout xml --layout-xml  layout.xml --gpx c.gpx nogopro.MP4 result.mov && mpv result.mov

real	0m13,390s
user	0m48,318s
sys	0m1,645s
```

### Using venv/bin/pip3.8 install gopro-overlay

I'm not sure about the arguments in this case? This is not working:

```bash
pkill ffmpeg ; pkill mpv ; time venv/bin/gopro-dashboard.py --use-gpx-only --overlay-size 1920x800 --video-time-end file-modified --generate overlay --profile overlay --layout xml --layout-xml  layout.xml --gpx c.gpx nogopro.MP4 result.mov && mpv result.mov
```



Sorry that I can't check it right now
[c.gpx.zip](https://github.com/time4tea/gopro-dashboard-overlay/files/9676482/c.gpx.zip)

**Reply by time4tea** [reply]
This is good info - I'll see what happens running it here....

**Comment by time4tea** [comment]
Ah - the "out of range" only appears with TZ set as UTC + xxx - So for me in "Europe/London" its all good, but for "Europe/Athens" it doesn't work. Thanks for reporting that.

**Reply by ptanov** [reply]
Nice catch :⁠-⁠)

**Comment by time4tea** [comment]
That issue is hopefully  fixed in 0.60.0

**Reply by ptanov** [reply]
I can confirm that 
Test 1) Generate overlay on top of the original video from gpx and non-gopro video (crop only segment that matches the video)
and
Test 2) Generate overlay without video (using GPX)

produced the same result - it works, thanks!

Why '--generate cannot be combined with --use-gpx-only', (so Test 3 - Generate overlay only for the part of the non-gopro video (crop only segment that matches the video) is not supported)?

**Reply by time4tea** [reply]
Yeah.. just didn't figure out what the best behaviour would be there. Some people want the whole gpx track to be available in the "journey" view...

**Comment by time4tea** [comment]
I realise you also asked another question there... so for `--gpx-only` i was thinking, if you don't supply a source video, then it implies that you want only the dashboard video, if you *do* supply a source video, then it implies that you want the dashboard overlaid on top of it. I thought that would be more "natural". Happy to make adjustments based on feedback!

**Reply by ptanov** [reply]
In case 3) user should be able to use video just to extract the correct portion of the gpx file. Imagine this case - I'm in a multi-day trek/ski touring, I have a GPS device that record the track during the day and non GoPro camera (or drone) which I use to record short videos. When I come back to my home I would like to create video that includes most of the recorded short videos with telemetry overlays. In order to combine them I use KDEnlive. I don't want to reencode my videos twice - once with gopro-dashboard (to put the overlay) and then with KDEenlive (that will cause quality degradation and it will be slower) so I want to generate just the overlay using gopro-dashboard and then I'll group the generated overlay with corresponding video in the KDEenlive timeline so I can modify (cut/crop) them in parallel and at the end exported video will be reencoded just once (it will be faster and with better quality).
Currently for every video I'll have to open GPX file with a text editor and to manually remove all <trkpt> records that don't belong to the video and to use `--use-gpx-only --overlay-size 1920x800 --profile overlay` without video in order to generate just the overlay. So if I can use `--video-time` with `--generate overlay` and non gopro video this will save me time by automating the manual work of editing GPX file to extract only track points during video recording.
Sorry I'm not sure if I explained this case well :(

**Comment by KyleGW** [comment]
This is working well for me. I actually ended up needing the functionality for my most recent video from this weekend. My main GoPro I use for the GPS track overheated and shut down a few times. If I had to use just that track, I would have had missing pieces of the flight track. 
I was also recording GPS on my Garmin watch, so I was able to use that to generate the overlay and throw it on top the video in my editor. Bonus - I am also now able to overlay heart rate and see it go up as I was practicing each landing.


![Screenshot 2022-10-05 192721 (Medium)](https://user-images.githubusercontent.com/9725098/194187442-92c41671-7112-4aec-bf4b-c70ff1a2b71f.jpg)
